File: src/lib/api/schemas/construction.ts
/**
 * Zod schemas for construction pipeline API responses.
 *
 * Validates raw JSON from /construction-pipeline/* endpoints
 * and transforms snake_case → camelCase.
 */
import { z } from 'zod';

// ---------- Project Record Schema ----------

export const projectRecordSchema = z
  .object({
    id: z.number(),
    project_name: z.string().nullable(),
    project_address: z.string().nullable(),
    city: z.string().nullable(),
    submarket_cluster: z.string().nullable(),
    pipeline_status: z.string().nullable(),
    primary_classification: z.string().nullable(),
    number_of_units: z.number().nullable(),
    number_of_stories: z.number().nullable(),
    year_built: z.number().nullable(),
    developer_name: z.string().nullable(),
    owner_name: z.string().nullable(),
    latitude: z.number().nullable(),
    longitude: z.number().nullable(),
    building_sf: z.number().nullable(),
    avg_unit_sf: z.number().nullable(),
    star_rating: z.string().nullable(),
    rent_type: z.string().nullable(),
    vacancy_pct: z.number().nullable(),
    estimated_delivery_date: z.string().nullable(),
    construction_begin: z.string().nullable(),
    for_sale_price: z.number().nullable(),
    source_type: z.string().nullable(),
  })
  .transform((r) => ({
    id: r.id,
    projectName: r.project_name,
    projectAddress: r.project_address,
    city: r.city,
    submarketCluster: r.submarket_cluster,
    pipelineStatus: r.pipeline_status,
    primaryClassification: r.primary_classification,
    numberOfUnits: r.number_of_units,
    numberOfStories: r.number_of_stories,
    yearBuilt: r.year_built,
    developerName: r.developer_name,
    ownerName: r.owner_name,
    latitude: r.latitude,
    longitude: r.longitude,
    buildingSf: r.building_sf,
    avgUnitSf: r.avg_unit_sf,
    starRating: r.star_rating,
    rentType: r.rent_type,
    vacancyPct: r.vacancy_pct,
    estimatedDeliveryDate: r.estimated_delivery_date,
    constructionBegin: r.construction_begin,
    forSalePrice: r.for_sale_price,
    sourceType: r.source_type,
  }));

// ---------- Paginated Response ----------

export const projectsResponseSchema = z
  .object({
    data: z.array(projectRecordSchema),
    total: z.number(),
    page: z.number(),
    page_size: z.number(),
    total_pages: z.number(),
  })
  .transform((r) => ({
    data: r.data,
    total: r.total,
    page: r.page,
    pageSize: r.page_size,
    totalPages: r.total_pages,
  }));

// ---------- Filter Options ----------

export const constructionFilterOptionsSchema = z
  .object({
    submarkets: z.array(z.string()),
    cities: z.array(z.string()),
    statuses: z.array(z.string()),
    classifications: z.array(z.string()),
    rent_types: z.array(z.string()),
  })
  .transform((r) => ({
    submarkets: r.submarkets,
    cities: r.cities,
    statuses: r.statuses,
    classifications: r.classifications,
    rentTypes: r.rent_types,
  }));

// ---------- Analytics Schemas ----------

export const pipelineSummaryItemSchema = z
  .object({
    status: z.string(),
    project_count: z.number(),
    total_units: z.number(),
  })
  .transform((r) => ({
    status: r.status,
    projectCount: r.project_count,
    totalUnits: r.total_units,
  }));

export const pipelineFunnelItemSchema = z
  .object({
    status: z.string(),
    project_count: z.number(),
    total_units: z.number(),
    cumulative_units: z.number(),
  })
  .transform((r) => ({
    status: r.status,
    projectCount: r.project_count,
    totalUnits: r.total_units,
    cumulativeUnits: r.cumulative_units,
  }));

export const permitTrendPointSchema = z
  .object({
    period: z.string(),
    source: z.string(),
    series_id: z.string(),
    value: z.number(),
  })
  .transform((r) => ({
    period: r.period,
    source: r.source,
    seriesId: r.series_id,
    value: r.value,
  }));

export const employmentPointSchema = z
  .object({
    period: z.string(),
    series_id: z.string(),
    value: z.number(),
  })
  .transform((r) => ({
    period: r.period,
    seriesId: r.series_id,
    value: r.value,
  }));

export const submarketPipelineItemSchema = z
  .object({
    submarket: z.string(),
    total_projects: z.number(),
    total_units: z.number(),
    proposed: z.number(),
    under_construction: z.number(),
    delivered: z.number(),
  })
  .transform((r) => ({
    submarket: r.submarket,
    totalProjects: r.total_projects,
    totalUnits: r.total_units,
    proposed: r.proposed,
    underConstruction: r.under_construction,
    delivered: r.delivered,
  }));

export const classificationBreakdownItemSchema = z
  .object({
    classification: z.string(),
    project_count: z.number(),
    total_units: z.number(),
  })
  .transform((r) => ({
    classification: r.classification,
    projectCount: r.project_count,
    totalUnits: r.total_units,
  }));

export const deliveryTimelineItemSchema = z
  .object({
    quarter: z.string(),
    total_units: z.number(),
    project_count: z.number(),
  })
  .transform((r) => ({
    quarter: r.quarter,
    totalUnits: r.total_units,
    projectCount: r.project_count,
  }));

export const constructionDataQualitySchema = z
  .object({
    total_projects: z.number(),
    projects_by_source: z.record(z.string(), z.number()),
    source_logs: z.array(z.record(z.string(), z.unknown())),
    null_rates: z.record(z.string(), z.number()),
    permit_data_count: z.number(),
    employment_data_count: z.number(),
  })
  .transform((r) => ({
    totalProjects: r.total_projects,
    projectsBySource: r.projects_by_source,
    sourceLogs: r.source_logs,
    nullRates: r.null_rates,
    permitDataCount: r.permit_data_count,
    employmentDataCount: r.employment_data_count,
  }));

export const constructionImportStatusSchema = z
  .object({
    unimported_files: z.array(z.string()),
    last_imported_file: z.string().nullable(),
    last_import_date: z.string().nullable(),
    total_projects: z.number(),
  })
  .transform((r) => ({
    unimportedFiles: r.unimported_files,
    lastImportedFile: r.last_imported_file,
    lastImportDate: r.last_import_date,
    totalProjects: r.total_projects,
  }));

export const triggerConstructionImportResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

File: backend/app/schemas/construction.py
"""
Construction pipeline schemas for API request/response validation.

Extracted from backend/app/api/v1/endpoints/construction_pipeline.py
to follow the project convention of schemas in app/schemas/.
"""

from datetime import date
from typing import Any

from .base import BaseSchema


class ProjectRecord(BaseSchema):
    """A single construction project record."""

    id: int
    project_name: str | None = None
    project_address: str | None = None
    city: str | None = None
    submarket_cluster: str | None = None
    pipeline_status: str | None = None
    primary_classification: str | None = None
    number_of_units: int | None = None
    number_of_stories: int | None = None
    year_built: int | None = None
    developer_name: str | None = None
    owner_name: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    building_sf: float | None = None
    avg_unit_sf: float | None = None
    star_rating: str | None = None
    rent_type: str | None = None
    vacancy_pct: float | None = None
    estimated_delivery_date: date | None = None
    construction_begin: str | None = None
    for_sale_price: float | None = None
    source_type: str | None = None


class PaginatedProjectsResponse(BaseSchema):
    """Paginated list of construction projects."""

    data: list[ProjectRecord]
    total: int
    page: int
    page_size: int
    total_pages: int


class ConstructionFilterOptionsResponse(BaseSchema):
    """Available filter values for construction pipeline dropdowns."""

    submarkets: list[str]
    cities: list[str]
    statuses: list[str]
    classifications: list[str]
    rent_types: list[str]


class PipelineSummaryItem(BaseSchema):
    """Summary of projects by pipeline status."""

    status: str
    project_count: int
    total_units: int


class PipelineFunnelItem(BaseSchema):
    """Funnel chart data point with cumulative units."""

    status: str
    project_count: int
    total_units: int
    cumulative_units: int


class PermitTrendPoint(BaseSchema):
    """Permit trend time-series data point."""

    period: str
    source: str
    series_id: str
    value: float


class EmploymentPoint(BaseSchema):
    """Employment time-series data point."""

    period: str
    series_id: str
    value: float


class PermitVelocityPoint(BaseSchema):
    """Permit velocity data point."""

    source: str
    period: str
    count: int
    total_value: float


class SubmarketPipelineItem(BaseSchema):
    """Pipeline breakdown by submarket."""

    submarket: str
    total_projects: int
    total_units: int
    proposed: int
    under_construction: int
    delivered: int


class ClassificationBreakdownItem(BaseSchema):
    """Pipeline breakdown by classification."""

    classification: str
    project_count: int
    total_units: int


class DeliveryTimelineItem(BaseSchema):
    """Delivery timeline data point by quarter."""

    quarter: str  # e.g. "Q1 2026"
    total_units: int
    project_count: int


class ConstructionDataQualityReport(BaseSchema):
    """Data quality report for construction pipeline."""

    total_projects: int
    projects_by_source: dict[str, int]
    source_logs: list[dict[str, Any]]
    null_rates: dict[str, float]
    permit_data_count: int
    employment_data_count: int


class ConstructionImportResponse(BaseSchema):
    """Response from construction data import."""

    success: bool
    message: str
    rows_imported: int = 0
    rows_updated: int = 0


class ConstructionImportStatusResponse(BaseSchema):
    """Status of construction data imports."""

    unimported_files: list[str]
    last_imported_file: str | None = None
    last_import_date: str | None = None
    total_projects: int = 0


class FetchAllResponse(BaseSchema):
    """Response from fetch-all data sources operation."""

    success: bool
    message: str
    results: dict[str, Any] = {}


class BackfillResponse(BaseSchema):
    """Response from data backfill operation."""

    success: bool
    message: str
    rows_updated: int = 0
